src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNNQuantizationStateReader.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.quantization.enums.ScalarQuantizationType;
+import org.opensearch.knn.quantization.models.quantizationParams.ScalarQuantizationParams;
+import org.opensearch.knn.quantization.models.quantizationState.MultiBitScalarQuantizationState;
+import org.opensearch.knn.quantization.models.quantizationState.OneBitScalarQuantizationState;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationStateReadConfig;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Reads quantization states
+ */
+public final class KNNQuantizationStateReader {
+
+    /**
+     * Read quantization states and return list of fieldNames and bytes
+     * File format:
+     * Header
+     * QS1 state bytes
+     * QS2 state bytes
+     * Number of quantization states
+     * QS1 field number
+     * QS1 state bytes length
+     * QS1 position of state bytes
+     * QS2 field number
+     * QS2 state bytes length
+     * QS2 position of state bytes
+     * Position of index section (where QS1 field name is located)
+     * -1 (marker)
+     * Footer
+     *
+     * @param state the read state to read from
+     */
+    public static Map<String, byte[]> read(SegmentReadState state) throws IOException {
+        String quantizationStateFileName = getQuantizationStateFileName(state);
+        Map<String, byte[]> readQuantizationStateInfos = new HashMap<>();
+
+        try (IndexInput input = state.directory.openInput(quantizationStateFileName, IOContext.READ)) {
+            CodecUtil.retrieveChecksum(input);
+
+            int numFields = getNumFields(input);
+
+            List<Integer> fieldNumbers = new ArrayList<>();
+            List<Long> positions = new ArrayList<>();
+            List<Integer> lengths = new ArrayList<>();
+
+            // Read each field's metadata from the index section
+            for (int i = 0; i < numFields; i++) {
+                fieldNumbers.add(input.readInt());
+                int length = input.readInt();
+                lengths.add(length);
+                long position = input.readVLong();
+                positions.add(position);
+            }
+            // Read each field's bytes
+            for (int i = 0; i < numFields; i++) {
+                byte[] stateBytes = readStateBytes(input, positions.get(i), lengths.get(i));
+                String fieldName = state.fieldInfos.fieldInfo(fieldNumbers.get(i)).getName();
+                readQuantizationStateInfos.put(fieldName, stateBytes);
+            }
```

Comment:
I don't think this works because we are reading all of the positions and lists back to back in an index section.  All of the bytes are in their own section.

Reply:
I misunderstood, will push change shortly to be more efficient.